// mendixmodelsdk/dist/sdk/internal/deltas/delta-reverser.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const units_1 = require("../units");
const deltaUtils = require("./delta-utils");
class DeltaReverser {
    constructor(model) {
        this.model = model;
    }
    reverse(delta) {
        const unit = deltaUtils.getUnitForDelta(this.model, delta);
        switch (delta.deltaType) {
            case "UPDATE_PROPERTY_VALUE":
                return this.reverseUpdatePropertyValueDelta(delta, deltaUtils.asModelUnit(unit));
            case "CREATE_ELEMENT_TREE":
                return this.reverseCreateElementTreeDelta(delta);
            case "DELETE_ELEMENT":
                return this.reverseDeleteElementDelta(delta, deltaUtils.asModelUnit(unit));
            case "ATTACH_ELEMENT":
                return this.reverseAttachElementDelta(delta);
            case "DETACH_ELEMENT":
                return this.reverseDetachElementDelta(delta, deltaUtils.asModelUnit(unit));
            default:
                throw new Error(`Delta type ${delta.deltaType} is not supported.`);
        }
    }
    reverseUpdatePropertyValueDelta(delta, unit) {
        const structure = (unit instanceof units_1.StructuralUnit) ?
            unit :
            deltaUtils.findElement(this.model, deltaUtils.asModelUnit(unit), delta.elementId);
        const property = deltaUtils.findProperty(structure, delta.propertyName);
        let reversedMutator;
        const mutator = delta.mutator;
        switch (mutator.mutatorType) {
            case "CHANGE":
                if (deltaUtils.isListProperty(property)) {
                    reversedMutator = {
                        mutatorType: "CHANGE",
                        value: deltaUtils.rawList(property)[mutator.updateIndex],
                        updateIndex: mutator.updateIndex
                    };
                }
                else {
                    reversedMutator = {
                        mutatorType: "CHANGE",
                        value: deltaUtils.rawValue(property)
                    };
                }
                break;
            case "ADD":
                reversedMutator = {
                    mutatorType: "REMOVE",
                    removalIndex: mutator.insertionIndex !== undefined ? mutator.insertionIndex : deltaUtils.rawList(property).length
                };
                break;
            case "REMOVE":
                reversedMutator = {
                    mutatorType: "ADD",
                    value: deltaUtils.rawList(property)[mutator.removalIndex],
                    insertionIndex: mutator.removalIndex
                };
                break;
            default:
                throw new Error(`Invalid mutatorType: ${JSON.stringify(mutator)}`);
        }
        return Object.assign({}, delta, { mutator: reversedMutator });
    }
    reverseCreateElementTreeDelta(delta) {
        return {
            deltaType: "DELETE_ELEMENT",
            unitId: delta.unitId,
            elementId: delta.elementTree.$ID
        };
    }
    reverseDeleteElementDelta(delta, unit) {
        const containmentInfo = deltaUtils.getContainmentInfo(unit, this.model, delta.elementId);
        return {
            deltaType: "CREATE_ELEMENT_TREE",
            unitId: delta.unitId,
            additionIndex: containmentInfo.index,
            parentId: containmentInfo.parentId,
            parentPropertyName: containmentInfo.parentPropertyName,
            elementTree: containmentInfo.element.toJSON()
        };
    }
    reverseAttachElementDelta(delta) {
        return {
            deltaType: "DETACH_ELEMENT",
            unitId: delta.unitId,
            elementId: delta.elementId
        };
    }
    reverseDetachElementDelta(delta, unit) {
        const containmentInfo = deltaUtils.getContainmentInfo(unit, this.model, delta.elementId);
        return {
            deltaType: "ATTACH_ELEMENT",
            unitId: delta.unitId,
            elementId: delta.elementId,
            newIndex: containmentInfo.index,
            newParentId: containmentInfo.parentId,
            newParentPropertyName: containmentInfo.parentPropertyName
        };
    }
}
exports.DeltaReverser = DeltaReverser;
