// mendixmodelsdk/dist/sdk/internal/transaction-checks.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const deltaUtils = require("./deltas/delta-utils");
const elements_1 = require("./elements");
class TransactionChecks {
    constructor(model) {
        this.model = model;
        this.checkDetachedStructures = true;
        this.detachedRequiredProperties = [];
    }
    check(queue) {
        if (this.checkDetachedStructures && this.model.detachedStructures.length > 0) {
            throw new Error("Some structures that were detached were not attached within the same operation");
        }
        for (const property of this.detachedRequiredProperties) {
            if (!property.get()) {
                throw new Error(`Required part property '${property.name}' has null value`);
            }
        }
        this.detachedRequiredProperties = [];
    }
    onNewDelta(delta) {
        if (delta.deltaType === "DETACH_ELEMENT") {
            const unit = deltaUtils.asModelUnit(deltaUtils.getUnitForDelta(this.model, delta));
            const element = deltaUtils.findElement(this.model, unit, delta.elementId);
            const handle = element.container._childHandle(element);
            if (handle.containingProperty.isRequired && element.container instanceof elements_1.AbstractElement) {
                this.detachedRequiredProperties.push(handle.containingProperty);
            }
        }
    }
}
exports.TransactionChecks = TransactionChecks;
