// mendixmodelsdk/dist/sdk/internal/units.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const elements_1 = require("./elements");
const structures_1 = require("./structures");
const instances_1 = require("./instances");
const properties_1 = require("./properties");
/**
 * Base class for implementations of {@link IAbstractUnit}.
 */
class AbstractUnit extends structures_1.Structure {
    constructor(model, structureTypeName, id, isPartial, container) {
        super(model, structureTypeName, id, isPartial, container);
        /** @internal */
        this._isLoading = false;
        /** @internal */
        this._afterLoadCallbacks = [];
    }
    /**
     * Checks whether all attributes are available at this instant;
     * if false, a load is required to access these properties.
     */
    get isLoaded() {
        return !this._isPartial;
    }
    /** @internal */
    _markNotLoaded() {
        this._isLoading = false;
        this._isPartial = true;
    }
    /** @internal */
    _markLoaded() {
        this._isLoading = false;
        this._isPartial = false;
        this._afterLoadCallbacks.splice(0).forEach(cb => cb(this));
    }
    /** @internal */
    _markLoading() {
        if (this._isLoading || this.isLoaded) {
            throw new Error("Illegal state: cannot re-load a unit that's already loaded");
        }
        this._isLoading = true;
    }
    /** @internal */
    _registerAfterLoadCallback(callback) {
        this._afterLoadCallbacks.push(callback);
    }
    /** @internal */
    _updateWithJsonImpl(json) {
        super._updateWithJsonImpl(json.contents);
    }
    /** @internal */
    _dispose() {
        this._model._removeUnit(this);
        super._dispose();
    }
    /**
     * Sends the create delta for this unit and all pending deltas.
     * @internal
     */
    _handleCreateSelf() {
        this.unit._model._sendDelta({
            deltaType: "CREATE_UNIT",
            unitId: this.id,
            containerId: this.container.id,
            containmentName: this._containmentName,
            contentType: this.structureTypeName
        });
        this._state = "attached";
        this._pendingDeltasAfterCreate.forEach(delta => {
            delta.unitId = this.unit.id;
            this.unit._model._sendDelta(delta);
        });
    }
    deepCopyInto(newParent) {
        throw new Error("deepCopy for AbstractUnit not implemented.");
    }
}
exports.AbstractUnit = AbstractUnit;
/**
 * Implementation of {@link IStructuralUnit}.
 */
class StructuralUnit extends AbstractUnit {
    constructor(model, structureTypeName, id, _ignoredIsPartial, container) {
        // _ignoredPartial argument is here, because structural units are always loaded completely,
        // but it is nice to keep the constructors for all units uniform so that the factories are simpler.
        super(model, structureTypeName, id, false, container);
    }
    get unit() {
        return this;
    }
    deepCopyInto(newParent) {
        // MWE: note that this is techinically possible, but what should be the semantics?
        // also clone structural children? Then the need to be loaded first. See also
        // comments in StructuralChild properties.
        throw new Error("Structural units like folders and modules cannot be cloned");
    }
    delete() {
        throw new Error("Deleting structural units is not supported yet");
        // TODO Create a delta to delete a structural unit and remove it from the qualified name cache
        // this._model._qualifiedNameCache.removeStructureFromCache(this);
        // Send delta
        // this._dispose();
    }
}
exports.StructuralUnit = StructuralUnit;
/**
 * Implementation of {@link IModelUnit}.
 */
class ModelUnit extends elements_1.AbstractElement {
    constructor(model, structureTypeName, id, isPartial, container) {
        super(model, structureTypeName, id, isPartial, container);
        /** @internal */
        this._isLoading = false;
        /* mimics multiple inheritance */
        /** @internal */
        this._markNotLoaded = AbstractUnit.prototype._markNotLoaded;
        /** @internal */
        this._markLoaded = AbstractUnit.prototype._markLoaded;
        /** @internal */
        this._markLoading = AbstractUnit.prototype._markLoading;
        /** @internal */
        this._registerAfterLoadCallback = AbstractUnit.prototype._registerAfterLoadCallback;
        /** @internal */
        // Note: flagged as unused by tslint but needs to be in (see failing unit tests otherwise)!
        // tslint:disable-next-line:no-unused-variable
        this._afterLoadCallbacks = [];
    }
    /** @internal */
    _handleCreateSelf() {
        AbstractUnit.prototype._handleCreateSelf.call(this);
        for (const property of this.allProperties()) {
            if (property instanceof properties_1.PartProperty) {
                const childElement = property.get();
                if (childElement) {
                    childElement._sendCreateDelta(this, property.name);
                }
            }
            else if (property instanceof properties_1.PartListProperty) {
                const childElements = property.get();
                for (let index = 0; index < childElements.length; index++) {
                    childElements[index]._sendCreateDelta(this, property.name, index);
                }
            }
        }
    }
    /** @internal */
    _updateWithJsonImpl(json) {
        AbstractUnit.prototype._updateWithJsonImpl.call(this, json); // delegate
    }
    get unit() {
        return this;
    }
    get isLoaded() {
        return !this._isPartial;
    }
    /** @internal */
    _dispose() {
        this._model._removeUnit(this);
        super._dispose(); // follow AbstractElement chain
    }
    /**
     * Given the type and name of an element, returns the element in this model unit
     * with the specified type and name if there is exactly one matching element, otherwise null.
     *
     * @internal
     */
    _resolveLocalName(elementTypeName, name) {
        if (!elementTypeName) {
            throw new Error("please provide an element type name");
        }
        const elementType = instances_1.instancehelpers.lookupClass(elementTypeName, this._model._allModelClasses());
        const matchingElements = [];
        this.traverse((abstractElement) => {
            if (abstractElement instanceof elementType && abstractElement.name === name) {
                matchingElements.push(abstractElement);
            }
        });
        return matchingElements.length === 1 ? matchingElements[0] : null;
    }
    /**
     * Given an element id, returns the element in this model unit with that id, otherwise null.
     */
    findElementById(id) {
        return this.traverseFind((abstractElement) => {
            return abstractElement.id === id ? abstractElement : null;
        });
    }
    /**
     * Deletes this element from the model.
     */
    delete() {
        if (this._isDeleted) {
            throw new Error("A deleted unit cannot be deleted");
        }
        const handle = this.container ? this.container._childHandle(this) : null;
        if (handle) {
            handle.delete();
        }
        else {
            this._deleteInternal();
        }
    }
    /** @internal */
    _deleteInternal() {
        // Make sure that the element does not delete twice.
        if (this._isDoingDelete) {
            return;
        }
        this._isDoingDelete = true;
        this._model._qualifiedNameCache.removeStructureFromCache(this);
        this._sendDeleteDelta();
        this._container = null;
        this.traverse(element => {
            element._state = "deleted";
        });
        this._dispose();
        this._isDoingDelete = false;
    }
}
exports.ModelUnit = ModelUnit;
